### DISCUSSION TITLE
Standardize syntax to execute terminal commands within .qmd?

### DISCUSSION THREAD

**Discussion by maxdrohde** [quarto-dev/quarto-cli#2648] [Q&A] [answered]
In a .qmd file using `knitr`, I can execute a terminal command like `ls` in a cell with

````
```{zsh}
ls
```
````

and in a .qmd file using `jupyter`, I can use

````
```{python}
!ls
```
````

Are there any plans to standardize this syntax? I would like to be able to use the knitr syntax in documents using the jupyter engine.

**Comment by rgaiacs** [comment, marked as answer]
Hi @maxdrohde,

Thanks for your contribution.

> Are there any plans to standardize this syntax?

Quarto syntax is standardize but knitr and jupyter work different and, unfortunately, you need to tailor your document to the computation engine that you will use.

# Long Explanation

knitr supports more than one language per document. For example, 

``````
```{r}
1 + 1
```

Foo

```{bash}
echo Awesome
```
``````

A list of all the languages supported by knitr is available at https://bookdown.org/yihui/rmarkdown-cookbook/other-languages.html.

Jupyter only support one language per document. For example,

``````
```{python}
1 + 1
```

Foo

```{bash}
echo Awesome
```
``````

is a valid Quarto document but it is a invalid Jupyter notebook because you are using Python and Bash in the same document.

The IPython kernel has a feature called "shell assignment with !". This is what you mention. My understanding is that it will not work if you are using a kernel other than the IPython.

Jupyter also has extra features, for example line magics (`%`) and cell magics (`%%`). They don't work with knitr.

> I would like to be able to use the knitr syntax in documents using the jupyter engine.

This is a feature request to Jupyter already register at https://github.com/jupyterlab/jupyterlab/issues/2815.

**Reply by maxdrohde** [reply]
Thank you for this great explanation!
